chain-crypto/src/algorithms/vrf/dleq.rs

```diff
@@ -0,0 +1,145 @@
+use crate::ec::ristretto255::{GroupElement, Scalar};
+use cryptoxide::blake2b::Blake2b;
+use cryptoxide::digest::Digest;
+
+/// Proof of discrete logarithm equivalence
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Proof {
+    c: Challenge,
+    z: Scalar,
+}
+
+impl Proof {
+    pub const PROOF_SIZE: usize = Scalar::BYTES_LEN * 2;
+
+    pub fn to_bytes(&self, output: &mut [u8]) {
+        assert_eq!(output.len(), Self::PROOF_SIZE);
+        output[0..Scalar::BYTES_LEN].copy_from_slice(&self.c.0.to_bytes());
+        output[Scalar::BYTES_LEN..].copy_from_slice(&self.z.to_bytes());
+    }
+
+    pub fn from_bytes(slice: &[u8]) -> Option<Self> {
+        if slice.len() != Self::PROOF_SIZE {
+            return None;
+        }
+        let mut c_array = [0u8; Scalar::BYTES_LEN];
+        c_array.copy_from_slice(&slice[0..Scalar::BYTES_LEN]);
+        let c = Scalar::from_bytes(&c_array)?;
+
+        let mut z_array = [0u8; Scalar::BYTES_LEN];
+        z_array.copy_from_slice(&slice[Scalar::BYTES_LEN..]);
+        let z = Scalar::from_bytes(&z_array)?;
+
+        let proof = Proof { c: Challenge(c), z };
+        Some(proof)
+    }
+}
+
+/// Parameters for DLEQ where g1^a = h1, h2^a = h2
+pub struct Dleq<'a> {
+    pub g1: &'a GroupElement,
+    pub h1: &'a GroupElement,
+    pub g2: &'a GroupElement,
+    pub h2: &'a GroupElement,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Challenge(Scalar);
+
+fn challenge(
+    h1: &GroupElement,
+    h2: &GroupElement,
+    a1: &GroupElement,
+    a2: &GroupElement,
+) -> Challenge {
```

Comment:
challenge should be computed using the two bases and the two points (`g1, g2, h1, h2`) together with the two announcements (`a1, a2`). I would keep the latest version of the ZKP we had, even if we repeat the code in `chain-vote`.